current/advanced/block-import.md

```diff
@@ -0,0 +1,41 @@
+---
+slug: block-import
+lang: en
+title: The Block Import Pipeline
+---
+
+
+## The Import Queue
+
+The import queue is an abstract worker queue present in every Substrate node. It is not part of the runtime. The Import Queue is responsible for processing pieces of incoming information, verifying them, and if they are valid, importing them into the node's state. The most fundamental piece of information that the import queue processes is blocks themselves, but it is also responsible for importing consensus-related messages such as Justifications and, in light clients, finality proofs.
+
+The Import queue collects incoming blocks from the network or from other parts of the local node, and stores them in a pool. The blocks are later checked for validity and discarded if they are not valid. Finally, the block is imported when the import queue is instructed to import it by the consensus engine.
```

Comment:
There are some things here that aren't entirely correct. On normal node operation the import queue will only deal with information coming from the network, blocks authored locally are imported directly without going through the import queue. I think the only situation where the import queue isn't used with network blocks is when doing a manual import `substrate import` in which the block data will be coming from a file and dumped into the queue. The queue runs on its own and is not "instructed by consensus", as blocks are pushed into the queue (typically from network sync), the queue background worker just imports them in-order (and if one fails to import it fails its subsequent descendents). The queue provides some hooks (`blocks_processed`, `justification_imported`, etc.) that can be used to follow the progress (usually by the sync module to adjust the sync algorithm). Additionally, other modules who are just interested in which blocks have been imported or finalized can just use `client.import_notification_stream()` or `finality_notification_stream()`.

The way that consensus logic hooks up into the import pipeline is through `BlockImport` which is used by the `ImportQueue`, i.e. the queue is calling into `BlockImport` whenever it tries to import a new block. This `BlockImport` instance can be a composition of multiple `BlockImport` stages that are aware of consensus logic (e.g. `GrandpaBlockImport(BabeBlockImport(ClientBlockImport))`, but it is still the import queue driving the process, not the other way around.

Reply:
What does the import queue architecture look like, is it a set of async workers acting on messages from the networking stack? Each provided with an instance of `BlockImport`? Or is it single threaded?
(I don't think that level of detail is relevant to this text, just a personal curiosity)

Reply:
That makes sense and sounds more consistent with David's expectations of what a queue was supposed to do. So, if not consensus engines, who calls methods like [`ImportQueue::import_blocks`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/trait.ImportQueue.html#tymethod.import_blocks)? Maybe the network calls them?